---
title: "Figure_2"
output: html_document
---

This markdown file contains the necessary code to to generate the data and diagrams for Figure 2 and S2
Several files generated in the markdown for Figure 1 are necessary to run these scripts

```{r distribution of motifs along the protein length}
#The position of the motifs is normalized to the protein length (0 = N-terminus, 1 = C-terminus)
getAbsFreq <- function(who, motif){
  subSet <- left_join(CMA.motifs, select(CMA.protRel,Entry,topGroups), by = "Entry")
  subSet <- eval(parse(text=paste0("subSet %>% filter(topGroups == \"",eval(who),"\", ",eval(motif)," == 1) ")))
  ggplot(subSet, aes(x= relPos)) + geom_histogram(breaks = seq(0,1,by = 0.02), size=0.5, boundary = 0, fill = "grey20")  + theme_classic() + scale_y_continuous(expand=c(0,0))  + theme(plot.title=element_text(size=9),axis.title.y=element_text(size=9),axis.title.x=element_text(size=9),axis.text.x=element_text(size=9, color="black"),axis.text.y=element_text(size=9,color="black"), axis.ticks = element_line(color = "black"), plot.margin  = unit(c(0.05,0.2,0.05,0.05),"inch"))
}

canonFreq <- getAbsFreq(who="canon", motif = "canonical")
canonFreq <- canonFreq +  theme(legend.position = "none") + labs(title="distribution of canonical motifs",x="relative protein length",y="n of motifs")   

phosFreq <- getAbsFreq(who="phos", motif = "phos")
phosFreq <- phosFreq +  theme(legend.position = "none") + labs(title="distribution of phosph.-gen. motifs",x="relative protein length",y="n of motifs") 

kFreq <- getAbsFreq(who="Konly", motif = "K") 
kFreq <- kFreq +  theme(legend.position = "none") + labs(title="distribution of acetyl.-gen. motifs",x="relative protein length",y="n of motifs")  

ggsave("images/canonical_density.png",canonFreq, height = 1.25, width = 2.75, device = "png")
ggsave("images/phos_density.png",phosFreq, height = 1.25, width = 2.75, device = "png")
ggsave("images/K_density.png",kFreq, height = 1.25, width = 2.75, device = "png")
```

To estimate the effect of the initiator Methionine the motif positions are calculated excluding the first N-terminal Methionine (column relPos_corr generated in for Fig1)

```{r omitting first Methionine}
getAbsFreq_noMeth <- function(who, motif){
  subSet <- left_join(CMA.motifs, select(CMA.protRel,Entry,topGroups), by = "Entry")
  subSet <- eval(parse(text=paste0("subSet %>% filter(topGroups == \"",eval(who),"\", ",eval(motif)," == 1) ")))
  ggplot(subSet, aes(x= relPos_corr)) + geom_histogram(breaks = seq(0,1,by = 0.02), size=0.5, boundary = 0, fill = "grey20")  + theme_classic() + scale_y_continuous(expand=c(0,0))  + theme(plot.title=element_text(size=9),axis.title.y=element_text(size=9),axis.title.x=element_text(size=9),axis.text.x=element_text(size=9, color="black"),axis.text.y=element_text(size=9,color="black"), axis.ticks = element_line(color = "black"), plot.margin  = unit(c(0.05,0.2,0.05,0.05),"inch"))
}

canonFreq_noMeth <- getAbsFreq_noMeth(who="canon", motif = "canonical")
canonFreq_noMeth <- canonFreq_noMeth +  theme(legend.position = "none") + labs(title="distribution of canonical motifs ",x="relative protein length",y="n of motifs")

ggsave("images/canonical_density_noMeth.png",canonFreq_noMeth, height = 1.25, width = 2.75, device = "png")

```

As a measure for the deacrease in density the mean number of motifs in the first 2.5% of the protein length versus remainder of the protein is calculated

```{r percent difference between N-term and rest of the protein}
getCvsRest <- function(who, motif ){
  subSet <- left_join(CMA.motifs, select(CMA.protRel,Entry,topGroups), by = "Entry")
  subSet <- eval(parse(text=paste0("subSet %>% filter(topGroups == \"",eval(who),"\", ",eval(motif)," == 1) ")))

  first2_5 <- subSet %>% filter(relPos <= 0.025) %>% summarise(n = n()) %>% mutate(n/2.5)
  rest <- subSet %>% filter(relPos > 0.025) %>% summarise(n = n()) %>% mutate(n/97.5)
  return(round(100*(first2_5/rest),digits=2))
}

getCvsRest(who="canon",motif="canonical")
```

For better visibility, the N- ad C-terminal regions are magnified

```{r fine grain histogram N-terminal vs C-terminal region}
AbsFreqTerm <- function(whichEnd){
  subSet <- left_join(CMA.motifs, select(CMA.protRel,Entry,topGroups), by = "Entry")
  subSet <- subSet %>% filter(topGroups == "canon", canonical == 1)
  if (whichEnd == "N-terminus")
    subSet <- subSet %>% filter(relPos <=0.1)
  else if (whichEnd == "C-terminus")
    subSet <- subSet %>% filter(relPos > 0.9)
  histogram <- ggplot(subSet,aes(x=relPos)) + geom_histogram(binwidth=0.001, boundary=0, fill="grey20") + theme_classic() + scale_y_continuous(expand=c(0,0)) + coord_cartesian(ylim = c(0,40)) + labs(title=paste0("region close to the ",whichEnd),x="relative amino acid position", y = "n of motifs") + theme(plot.title=element_text(size=9), axis.title.x=element_text(size=9) , axis.title.y=element_text(size=9), axis.text=element_text(color="black"), axis.ticks = element_line(color="black"))
  if (whichEnd == "C-terminus")
    histogram <- histogram + scale_x_reverse() + labs(title = "region close to the C-terminus (reversed)")
  return(histogram)
}

ggsave("images/canonical_nHisto.png", AbsFreqTerm("N-terminus"), height = 1.75, width = 3, device = "png")
ggsave("images/canonical_cHisto.png", AbsFreqTerm("C-terminus"), height = 1.75, width = 3, device = "png")
```

Secondary structure prediction is carried out using JPred4 batch submission. For batch submission the "massSubmitScheduler.csh" script from the JPred4 home page was used.

A list of experimentally described motifs (n = 28) was compiled from literature (Supplemental Table 1).  
Because there is an upper limit (800 AA) for the protein size to be predicted in JPred4, proteins with a Length > 800 amino acids (n = 4) were excluded from the analysis to avoid error introduced by manual truncation of the sequence.

Likewise for all proteins with one canonical motif in the reviewed human proteome (n = 1205) secondary structure was predicted using JPred4.  
Again proteins with a length > 800 aa were excluded from the analysis (n = 44). Additinally, JPred4 predicition failed for 14 additional proteins. The remaining 1147 proteins were used to quantify the percentage of exposed amino acids in the motif region 

The JPred predicition files (.jnet) are are loaded into R and parsed to extract the exposure information.  
The jnet files are available in raw_data as tarball archivess.

```{r getting the JPred predictions}
# Parsing the JPred4 output
getJPredExposure <- function(protein, indices = "one"){
  predSurf <- read.table(paste0("JPred4/one_canon_output_jnet/",protein,".jnet"),sep=",", stringsAsFactors = F, nrows=14)
  mixedCol <- predSurf[,1]
  predSurf[,1] <- sapply(strsplit(mixedCol,":"),"[",2)
  predSurf <- as.data.frame(t(predSurf[,c(1:ncol(predSurf)-1)]))
  colnames(predSurf) <- sapply(strsplit(mixedCol,":"),"[",1)
  # In case of the knonw substrates the indeces may have to be adjusted because of the 800 AA limit for JPred. Some proteins may be shorter and the indices have to be adjusted accordingly
  if (indices != "one"){
    # The start column starts with one for the string splitting
    predSurf$index <- c(seq_len(nrow(predSurf))+combined.motifs.JPred$start-1)
  } 
  else
    predSurf$index <- seq_len(nrow(predSurf))  
  predSurf$Entry <- protein
  return(select(predSurf, Entry, JNETSOL25, index))
}
```

Getting the predictions for the proteins with one canonical motif and plotting the results

```{r load files}
one_canon_jnet_files <- data.table::fread("raw_data/one_canon_output_jnet_files.txt", header = F)
colnames(one_canon_jnet_files) <- "filename"
one_canon_jnet_files$Entry <- sapply(strsplit(one_canon_jnet_files$filename,"\\."),"[",1)

oneCanon.JPred <- bind_rows(lapply(one_canon_jnet_files$Entry, getJPredExposure))

oneCanon.JPred$Score <- ifelse(oneCanon.JPred$JNETSOL25 == "B", 0, 1)
```

```{r preparing plot data}
# The data for the proteins with one canonical motif is split over different data sets
getRegion <- function(x, dataSet, motifData){
  
  motifPos <- motifData$motifPos[x]
  length <- motifData$Length[x]
  uniqueID <- motifData$uniqueID[x]
  
  boundaries <- c(max(1, motifPos - 30), min(motifPos + 30, length))
  dataSet <- dataSet %>% slice(boundaries[1]:boundaries[2])
  if (motifPos - 30 < 1){
    buffer <- 30 - motifPos + 1
    negRows <- data.frame(Entry = rep(dataSet$Entry[1], times = buffer), JNETSOL25 = NA, index = NA, Score = NA)
    dataSet <- rbind(negRows, dataSet)
  }
  if (motifPos + 30 > length){
    buffer <- 30 - length + motifPos
    posRows <- data.frame(Entry = rep(dataSet$Entry[1], times = buffer), JNETSOL25 = NA, index = NA, Score = NA)
    dataSet <- rbind(dataSet, posRows)
  }
  
  dataSet.t <- as.data.frame(t(dataSet[,4]))
  colnames(dataSet.t) <- seq(-30, 30, by = 1)
  dataSet.t$Entry <- dataSet$Entry[1]
  dataSet.t$uniqueID <- uniqueID
  return(dataSet.t)
}

makeExposurePlot <- function(Entry, dataSet, motifRef){
  expData <- eval(parse(text = paste0("dataSet %>% filter(Entry == \"",eval(Entry),"\")")))
  motifs <- eval(parse(text = paste0("motifRef %>% filter(Entry == \"",eval(Entry),"\")")))
  meanExp <- mean(expData$Score)
  print(nrow(motifs))
  motifRegions <- bind_rows(lapply(seq_len(nrow(motifs)), getRegion, dataSet = expData, motifData = motifs))
  motifRegions$meanExp <- meanExp
  return(motifRegions)
}
```

getting the plot data for the proteins with one canonical motif

```{r isolating region info}
# The motif ID column has to be added to make this compatible with the function for the known substrates
querySet$uniqueID <- paste0("oneCanon",seq_len(nrow(querySet)))

oneCanon.Regions <- bind_rows(lapply(one_canon_jnet_files$Entry, makeExposurePlot, dataSet = oneCanon.JPred, motifRef = querySet))
```

plotting the results as a bar plot of the percentage exposed with the mean percentage as a line

```{r plotting results}
makeExpPlot <- function(dataSet){
  plotData <- dataSet %>% summarise_at(vars(1:61), funs(sum = sum(., na.rm = T)))
  plotData.n <- dataSet %>% summarise_at(vars(1:61), funs(n = sum(!is.na(.)))) %>% unlist(.)

  plotData[1,] <- round(100*(plotData[1,] / plotData.n), digits = 2)
  plotData <- as.data.frame(t(plotData))
  colnames(plotData) <- "perc_exposed"
  plotData$position <- seq(-30, 30, by = 1)
  plotData$motif <- c(rep(0, times = 28), rep(1,times = 5), rep(0, times = 28))

  ggplot(plotData, aes(x= position, y = perc_exposed, color = motif)) + geom_bar(stat = "identity", show.legend = F) + geom_hline(yintercept = 100*(mean(oneCanon.Regions$meanExp)), color = "red") + theme_classic() + scale_y_continuous(expand=c(0,0)) + labs(x = "position relative to the middle (0) of the motifs", y="% exposed residues")  + theme(plot.title=element_text(size=10),axis.title.y=element_text(size=9),axis.title.x=element_text(size=9),axis.text.x=element_text(size=9, color="black"),axis.text.y=element_text(size=9,color="black")) 
}

ggsave("one_canon_exposure_plot.pdf",makeExpPlot(oneCanon.Regions), height = 1.75, width = 3)
```

Getting predictions for known substrates

The functions are slightly different because they allow for truncated sequences. This functionality was not used for the final analyses for reason described above but the functionality is preserved here for documentation.

```{r getting the JPred predictions}
getJPredExposure_ks <- function(protein, indices = "one"){
  predSurf <- read.table(paste0("raw_data/known_substrates_output_jnet/",protein,".jnet"),sep=",", stringsAsFactors = F, nrows=14)
  mixedCol <- predSurf[,1]
  predSurf[,1] <- sapply(strsplit(mixedCol,":"),"[",2)
  predSurf <- as.data.frame(t(predSurf[,c(1:ncol(predSurf)-1)]))
  colnames(predSurf) <- sapply(strsplit(mixedCol,":"),"[",1)
  # In case of the knonw substrates the indeces may have to be adjusted because of the 800 AA limit for JPred. Some proteins may be shorter and the indices have to be adjusted accordingly
  start <- eval(parse(text = paste0("known.motifs.JPred %>% filter(Entry == \"",eval(protein),"\") %>% select(start) %>% unlist(.)")))
  start <- as.integer(start[1])
  predSurf$index <- c(seq_len(nrow(predSurf))+start-1)
   
  predSurf$Entry <- protein
  return(select(predSurf, Entry, JNETSOL25, index))
}
```

```{r load JPred files}
known_jnet_files <- data.table::fread("JPred4/known_substrates_output_jnet_files.txt", header = F)
colnames(known_jnet_files) <- "filename"
known_jnet_files$Entry <- sapply(strsplit(known_jnet_files$filename,"\\."),"[",1)

# the motif information is loaded from file
known.motifs.JPred <- read.csv("known_substrates_motifs.csv", stringsAsFactors = F)

known.JPred <- bind_rows(lapply(known_jnet_files$Entry, getJPredExposure_ks))

known.JPred$Score <- ifelse(known.JPred$JNETSOL25 == "B", 0, 1)
```

```{r isolating region info}
Entries <- known.JPred %>% filter(!duplicated(Entry)) %>% select(Entry) %>% unlist(.)

known.motifs.filt.regions <- bind_rows(lapply(Entries, makeExposurePlot, dataSet = known.JPred, motifRef = known.motifs.JPred))
```

```{r}
ggsave("known_subs_noTrunc_exposure_plot.pdf",makeExpPlot(known.motifs.filt.regions), height = 1.75, width = 3)
```

Experimentally confirmed motifs are identified from literature and their sequence, pdb structures, JPred4 predicted structures, IUPred disorder scores are collected

This was later simplified to only include the pdb and JPred predicition data

```{r plotting disorder in region versus the whole protein}
#The sequence and position for experimentally confirmed motifs in 17 proteins are collected in the file "knownSubstrates_motifs.csv"
#Disorder scores are calculated using IUPred as described above

knownSubsMotifs <- read.csv("raw_data/conf_motifs/knownSubstrates_motifs.csv", stringsAsFactors = F)
refAAsize <- read.csv("raw_data/conf_motifs/amino_acid_surfaceArea_NEW.csv")

makeKnownMotifPlot <- function(x){
  protein <- knownSubsMotifs$ID[x]
  name <- knownSubsMotifs$name[x]
  indexMod <- knownSubsMotifs$indexMod[x]
  
  #load data for the protein (IUPred, exposurePDB, exposureJPred, motif positions)
IUPred <- read.table(paste0("raw_data/conf_motifs/IUPred/",protein,".txt-disorder.txt"),stringsAsFactors = F, skip = 9)
  colnames(IUPred) <- c("index","residue","disorderScore")

#The JPred output format contains some additional formatting and has to be cleand up before use
  predSurf <- read.table(paste0("raw_data/conf_motifs/JPred/",protein,"_JPred.txt"),sep=",", stringsAsFactors = F, nrows=14)
  mixedCol <- predSurf[,1]
  predSurf[,1] <- sapply(strsplit(mixedCol,":"),"[",2)
  predSurf <- as.data.frame(t(predSurf[,c(1:ncol(predSurf)-1)]))
  colnames(predSurf) <- sapply(strsplit(mixedCol,":"),"[",1)
  #JPred only accepts inputs up to 600 amino acids. For longer proteins a 600 amino acid fragment containing the motif position was chosen.
  #To properly align the JPred fragments to the other data the starting index has to be shifted by indexMod
  predSurf$index <- c(seq_len(nrow(predSurf))+indexMod)

  #adding the structural information from pdb records
  if (file.exists(paste0("raw_data/conf_motifs/pdb/pdb_",protein,".ent.txt"))){
    pdb <- read.pdb(paste0("raw_data/conf_motifs/pdb/pdb_",protein,".ent.txt"))
    sse <- dssp(pdb)
    pdbSurf <- data.frame(index = pdb$atom$resno[pdb$atom$elety=="CA"], threeLetter = pdb$atom$resid[pdb$atom$elety=="CA"], acc = sse$acc, stringsAsFactors = F)
    pdbSurf <- left_join(pdbSurf, refAAsize, by = "threeLetter")
    pdbSurf <- pdbSurf %>% mutate(percExp = acc/TienTheoretical)
    pdbSurf$pdbSurf <- ifelse(pdbSurf$percExp < 0.25,"B","-")
  } else{
    pdbSurf <- data.frame(index = seq_len(nrow(IUPred)), pdbSurf = rep("x",nrow(IUPred)))
  }

  #combining IUPred, JPred and pdb data sets and dealing with missing values
  plotData <- left_join(IUPred,select(pdbSurf,index,pdbSurf),by="index")
  plotData$pdbSurf[is.na(plotData$pdbSurf)] <- "x"
  plotData <- left_join(plotData,select(predSurf,index,predSurf=JNETSOL25),by="index")
  plotData$predSurf[is.na(plotData$predSurf)] <- "x"

  
  
  #extracting the information about the motifs
descMotifs <- data.frame(raw = unlist(strsplit(knownSubsMotifs$described_motif[x],";")),stringsAsFactors = F)
  descMotifs <- descMotifs %>% rowwise() %>% do(motif = unlist(strsplit(.$raw,"_"))[1], kind = unlist(strsplit(.$raw,"_"))[2], pos = unlist(strsplit(.$raw,"_"))[3])
  descMotifs <- as.data.frame(descMotifs)
  descMotifs <- descMotifs %>% mutate_all(funs(as.character(.)))
  
  #To diplay the data a base plot is generated with the IUPred, JPred and pdb information and the motif poitions are overlayed on this plot
  
  basePlot <- ggplot(plotData, aes(x=index,y=disorderScore)) + geom_line(size=0.4) + geom_hline(aes(yintercept=0.4),color="red", size = 0.3) + geom_hline(aes(yintercept=0),color="black",size=0.4) + geom_point(aes(x=index,y=-0.25,fill = pdbSurf),shape=22, size =2,stroke =0) + geom_point(aes(x=index,y=-0.45,fill = predSurf),shape=22, size = 2, stroke=0) + theme_classic() + scale_fill_manual(values=c("grey25","green3","white"),breaks=c("-","B"),labels=c("exposed","buried")) + scale_x_continuous(expand=c(0,0)) + scale_y_continuous(expand=c(0,0), breaks = c(-0.45, -0.25, 0,0.2,0.4,0.6,0.8,1), labels = c("jPred4", "pdb","0","0.2","0.4","0.6","0.8","1")) + coord_cartesian(ylim=c(-0.6,1)) + labs(x = "amino acid position", y = "disorder score") + theme(plot.title=element_text(size=9, hjust=0.5),axis.title.y=element_text(size=9,hjust=0.75),axis.title.x=element_text(size=9) ,axis.text.x=element_text(size=9, color="black"),axis.text.y=element_text(size=9,color="black"), axis.line = element_line(size = 0.4), axis.ticks = element_line(color="black") , legend.text =  element_text(size=9,color="black"), legend.background = element_rect(fill = NA) , legend.position = c(0.5,0.325), legend.title = element_blank(), legend.direction = "horizontal")
  
  motifKind <- factor(descMotifs$kind)
  colorVector <- c(rep("black",length(levels(motifKind))))
  colorVector[levels(motifKind)=="C"] <- "yellow2"
  colorVector[levels(motifKind)=="P"] <- "deepskyblue3"
  colorVector[levels(motifKind)=="K"] <- "chartreuse3"
  colorVector[levels(motifKind)=="N"] <- "orange"
  
  ggsave(paste0("images/paper_",protein,"_descMotifs.png"),basePlot + geom_vline(data=descMotifs, aes(xintercept=as.integer(pos), color = kind), show.legend = F) + scale_color_manual(values = colorVector) + labs(title = name), width =2.5, height = 2.2, device ="png" )
  
}

for(i in nrow(knownSubsMotifs))
  makeKnownMotifPlot(i)
```




